---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
```{r knitr,eval=F,echo=F}
library(rmarkdown)
library(knitrBootstrap)
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=9,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render("~/repos/pcbc_c4_analysis/code/Rmd/CountMatrixmiRNA_seqbuster.Rmd", output_dir = "~/Dropbox/Public/hsph/pcbc/miRNA_analysis_seqbuster.html")
render("~/repos/pcbc_c4_analysis/code/Rmd/CountMatrixmiRNA_seqbuster.Rmd", output_dir = "~/Dropbox/Public/hsph/pcbc/miRNA_analysis_seqbuster.md", output_format="md_document")
```

# metadata and covariates

```{r meta}
library(tidyr)
library(dplyr)
library(knitr)
library(synapseClient)
library(DESeq2)
library(edgeR)
library(vcd)
library(limma)
library(reshape2)
library(gplots)
library(ggplot2)
library(RColorBrewer)
library(corrplot)
library(psych)
library(pheatmap)

source("~/repos/pcbc_c4_analysis/code/R/lib/qc.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/covariateAnalysis.R")

synapseLogin()

# Get metadata
metadata_id = "syn3219876"
metadata_obj = synTableQuery(paste('SELECT * FROM',metadata_id,sep=' '))
metadata = metadata_obj@values %>% distinct(UID)
row.names(metadata) = metadata$UID

#mat_id = "syn3355993"
#mat_obj =  synGet(id=mat_id)
#mat = mat_obj@filePath
mat = read.csv("~/repos/pcbc_c4_analysis/data/mirna.csv", check.names = FALSE)
reads = read.table("~/repos/pcbc_c4_analysis/data/smallRNA_reads.tsv", check.names = FALSE, row.names=1, header=T)
keep = ( (reads %>% mutate(samples=rownames(reads)) %>% filter(mirna > 500000 & norm > 0.2))[,"samples"] )
metadata$size= reads[as.character(row.names(metadata)),"norm"]
metadata$size_cat= reads[as.character(row.names(metadata)),"breaks"]
metadata$size_cat= reads[as.character(row.names(metadata)),"breaks"]
# join meso-15 and meso-30
stages =  gsub("-5","Early",gsub("-30","Late",gsub("-15","Late",metadata$Diffname_short)))
metadata$Diffname_short = as.character(stages)
```

## Clean and sync tables

Just make sure the tables are sync, same order for samples

```{r clean_metadata}
meta_clean = metadata[colnames(mat), ]
meta_clean[meta_clean == 'N/A'] = NA
meta_clean[is.na(meta_clean)] = 'unk'
meta_clean[meta_clean == ''] = 'unk'
```

```{r clean}
meta_clean = meta_clean[keep,]
mat = mat[,keep]

keep_meta = meta_clean$Diffname_short != "unk"
meta_clean = meta_clean[keep_meta,]
mat = mat[,keep_meta]

meta_clean$mirna= log10(colSums(mat))
```


## Correlation of metadata variables

I looked for a way to get a correlation p-value and a rho value for categorical
variables. I found that, that is quite similar, and could give us an idea.

```{r cor}
cols = c("Originating_Lab_ID","High_Confidence_Donor_ID","DatemiRNAextracted", "Donor_Life_Stage",
         "Diffname_short","Gender","Tissue_of_Origin", "run", "lane", "index",
         "Cell_Type_of_Origin","Cell_Line_Type","Reprogramming_Gene_Combination","size_cat")
cols_format = sub("[::_::]","\n",cols) # reduce space of labels in heatpmap

comp = expand.grid(cols,cols)
c = apply(comp,1,corr_two_nominal_vars,meta_clean)
res = cbind(comp,do.call(rbind,c))
names(res)[3:4] = c("cor","pvalue")

cor_mat = tidyr::spread(res[,c(1,2,3)],"Var1","cor")
cor_mat = cor_mat[,2:ncol(cor_mat)]
cor_mat = matrix(unlist(cor_mat),nrow=length(cols))
row.names(cor_mat) = cols_format
colnames(cor_mat) = cols_format

pval_mat = tidyr::spread(res[,c(1,2,4)],"Var1","pvalue")
pval_mat = pval_mat[,2:ncol(pval_mat)]
pval_mat = matrix(unlist(pval_mat), nrow=length(cols))
row.names(pval_mat) = cols_format
colnames(pval_mat) = cols_format

cor_mat[is.na(cor_mat)] = 0
par(cex=0.7)
corrplot(cor_mat, p.mat = pval_mat, method = 'number', insig = "blank",  
         is.corr=FALSE, sig.level = 0.05,tl.cex = 0.9, cl.cex=0.9)
```

i would say that **Tissue_of_origin**, **run**, **size** and **Gender** are a linear convination of all
the rest. Trying to reduce the number of covariates that correlates ones with another.

## most-variable genes

```{r limma, cache=TRUE}
voom_exp = voom(mat[rowMeans(mat)>5,], design=NULL, plot=T)
clean_names = paste0(summarydata[colnames(voom_exp$E),"group"],"-",1:ncol(voom_exp$E))
tmp_mat = voom_exp$E
colnames(tmp_mat) = clean_names
# write.table(tmp_mat,"report/voom_counts.tsv", quote=F,sep="\t")
```


Just the heapmap with top-variable genes but using linear regression instead of 
rowsVar() function.

```{r most-var}
select = most_variable(voom_exp$E)

# The palette with grey:
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

difname_levels = unique(meta_clean$Diffname_short)
de_cols = cbPalette[1:length(difname_levels)]
names(de_cols) = difname_levels


pheatmap(voom_exp[select,], annotation = meta_clean[,c("Diffname_short","size", "Donor_Life_Stage","Tissue_of_Origin","Gender")], annotation_colors = list("Diffname_short" = de_cols),show_rownames = F, show_colnames = F)

mds(voom_exp$E[select,],condition = meta_clean$Diffname_short)
```


## DESeq2

### PCA

There are some values that correlates, with PCA should be easy to see if samples
have some bias to the previous variables. If there is any, this should be taking
into account during DE.

I used DESeq2 because for miRNA, I always found to normalized better, since there is quite variation in the top expressed miRNA, and many times bias to library preparation.

If we want to look for a general UP/DOWN regulation, this is not valid.

I tried many different variables, to make sure there wasn't a bias.

I expected some bias due to Donor.life.Stage, although blastocyst are mixed
with adult and fetal samples. 

It seems are separated by differentiation stage, but don't know if it makes sense.

```{r plotpca,eval=F}
dds = DESeqDataSetFromMatrix(mat[rowMeans(mat)>5,], colData = meta_clean, design =~1)
rlogMat = rlog(dds, blind= TRUE)

DESeq2::plotPCA(rlogMat, intgroup = "Donor_Life_Stage", ntop=200)
DESeq2::plotPCA(rlogMat, intgroup = "Diffname_short", ntop=200)
```

## covariate analysis

i will just remove miRNA that are not express in less than 10% of the samples.

```{r define}
# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','Cell_Type_of_Origin','Tissue_of_Origin',
                     'Reprogramming_Gene_Combination','Donor_Life_Stage',
                     'Originating_Lab','Gender','High_Confidence_Donor_ID')
#ContCovariates = c('numberReads','mappedReads')


PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(mat, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.1)

PROCESSED_COUNTS$plotHist

COVARIATES = meta_clean[,c(FactorCovariates, "size")]

```

### correlation of expression PCs to covariates without design

```{r exp-cov}
# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','Cell_Type_of_Origin','Tissue_of_Origin',
                     'Donor_Life_Stage','Originating_Lab','Gender')

# Initial normalisation usign voom (with NULL desing)
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)

# Find PC of gene expression and significant covariates that are highly correlated with PCs
DM = getDesignMatrix(meta_clean[,c(FactorCovariates, "size")],FactorCovariates)

preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DM$design, 'NULL design(voom-normalized)',isKeyPlot=TRUE)
preAdjustedSigCovars$PC_res[[1]]$plotData + theme(axis.text.x=element_text(size=10))
```

### correlation of weight PCs to covariates without design

```{r weight}
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$weight, DM$design, 'NULL design(voom-normalized)',isKeyPlot=TRUE)
preAdjustedSigCovars$PC_res[[1]]$plotData + theme(axis.text.x=element_text(size=10))

```

### correlation of weight PCs to covariates with design

The expression matrix is the same if you change the design, so I am using here only 
weight to see if something change or not.

```{r with-design}
# Re-normalise with COVARIATES desing matrix
#new_cov = c('Diffname_short','run','Tissue_of_Origin','Donor_Life_Stage','Gender','size','Originating_Lab')
# DM = getDesignMatrix(meta_clean[,new_cov],new_cov)
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM$design, plot=T)
#duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION,DM$design,block=meta_clean$Diffname_short)
# Find PC of re-adjusted gene expression and significant covariates that are highly correlated with PCs
postAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.ADJUSTED.GENE_EXPRESSION$weight, DM$design, 'all adjusted design(voom-normalized)',isKeyPlot=TRUE)
postAdjustedSigCovars$PC_res[[1]]$plotData + theme(axis.text.x=element_text(size=10))
```

### correlation of residuals PCs to covariates with design

There is no correlation with residual, what it means that there is not extra 
variation in the model?

```{r cor-res}
# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
fit = lapply( 1:nrow(VOOM.ADJUSTED.GENE_EXPRESSION$E), function(x){
    f = lm(VOOM.ADJUSTED.GENE_EXPRESSION$E[x,] ~ 0 + DM$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weight[x,])
    f$residuals
})

# result.lm = glm(t(VOOM.ADJUSTED.GENE_EXPRESSION$E) ~ 0 + DM$design, weights = t(VOOM.ADJUSTED.GENE_EXPRESSION$weight))

resAdjustedSigCovars = runPCAandPlotCorrelations(do.call(rbind,fit), DM$design, 'residual',isKeyPlot=TRUE)

resAdjustedSigCovars$PC_res[[2]]$plotData
```

### correlation of residuals PCs to covariates with design without group of interest


```{r cor-res-without-group}
reduced_FactorCovariates = c('run','Tissue_of_Origin',
                             'Originating_Lab')

# Find PC of gene expression and significant covariates that are highly correlated with PCs
DM = getDesignMatrix(meta_clean[,c(reduced_FactorCovariates, "size")],FactorCovariates)
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM$design, plot=T)


# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
fit = lapply( 1:nrow(VOOM.ADJUSTED.GENE_EXPRESSION$E), function(x){
    f = lm(VOOM.ADJUSTED.GENE_EXPRESSION$E[x,] ~ 0 + DM$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weight[x,])
    f$residuals
})

DM = getDesignMatrix(meta_clean[,c(FactorCovariates, "size")],FactorCovariates)

resAdjustedSigCovars = runPCAandPlotCorrelations(do.call(rbind,fit), DM$design, 'residual',isKeyPlot=TRUE)

resAdjustedSigCovars$PC_res[[2]]$plotData
```

## clustering of with basic filtering

i wanted to see if we use filters like remove miRNAs with expression <= X in % of samples, how
PCA/MDS changes

```{r play-clustering}
#voom_mat = voom( ), design=design, plot=T)
qc_figures =  function(m, i, c){
    d = data.frame(row.names = colnames(m), cov=c)
    #dds = DESeqDataSetFromMatrix(mat_norm,colData = as.data.frame(design),design = ~ 1)
    #dds = estimateSizeFactors(dds)
    #size_factor = sizeFactors(dds)
    voom_mat = voom(m, design = model.matrix(~ 0 + c), plot=T)
    corrected = voom_mat$E / voom_mat$weight
    
    x = melt(voom_mat$E)
    idx = match(x$X2, i$UID)
    x$type = i[idx,"Diffname_short"]
    
    g1 = ggplot(x, aes(x=value,group=X2,colour=type)) + geom_line(aes(color=type), stat="density",  alpha=0.4) + xlab("expression")
    g2 = ggplot(x, aes(x=X2,y=value,colour=type)) + geom_jitter(size=1,alpha=0.3)+theme_bw() + xlab("")
    
    g3 = mds(voom_mat$weight, condition = i$Diffname_short, d = "cor", xi = 1,yi = 2) +ggtitle("PCA of voom_weights")
    # mds(voom_mat$weight, condition = meta_clean_3states$run, d = "cor", xi = 1,yi = 2)
    E_empty = voom_mat$E
    colnames(E_empty) = 1:ncol(E_empty)
    g4 = mds(E_empty, condition = i$Diffname_short, d = "cor", xi = 1,yi = 2)+ggtitle("PCA of  log2(voom_counts)")
  
    print(gridExtra::arrangeGrob(g1,g2,g3,g4))
}

meta_sub = meta_clean %>% filter(Diffname_short %in% c("MESOEarly", "SC", "DE", "EB", "ECTO"))
table(meta_sub$Diffname_short)

mat_sub= mat[,meta_sub$UID]

new_factors = c("Tissue_of_Origin","run","Originating_Lab_ID")
covariates = do.call(paste0,meta_sub[,new_factors])
```

### removing mirnas with less than 10 counts in less than 10 samples

```{r play-10-10,results='asis'}
mat_clean =  (mat_sub[rowSums(mat_sub>10)>10,])
cat(paste("number of miRNA", nrow(mat_clean)))
#mat_norm = t(apply(mat_clean, 1, winsorize, fraction=0.3)) 
mat_norm = mat_clean
colnames(mat_norm) = colnames(mat_sub)
qc_figures(mat_norm, meta_sub, covariates)

voom_mat = voom(mat_norm, design = model.matrix(~ 0 + ., model.frame(~ 0 + . , meta_sub[,c(new_factors, "Diffname_short")])), plot=T)
pheatmap(voom_mat$E, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type"),drop=F], show_rownames = F)
```

### removing mirnas with less than 10 counts in less than 60 samples

```{r play-60-10,results='asis'}
mat_clean =  (mat_sub[rowSums(mat_sub>10)>60,])
cat(paste("number of miRNA", nrow(mat_clean)))
#mat_norm = t(apply(mat_clean, 1, winsorize, fraction=0.3)) 
mat_norm = mat_clean
colnames(mat_norm) = colnames(mat_sub)
qc_figures(mat_norm, meta_sub, covariates)

voom_mat = voom(mat_norm, design = model.matrix(~ 0 + ., model.frame(~ 0 + . , meta_sub[,c(new_factors, "Diffname_short")])), plot=T)
pheatmap(voom_mat$E, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type"),drop=F], show_rownames = F)

```

### Filtering by group: SC data

I want to try if filtering by stable miRNAs give better signal. 
I keep miRNAs that are express >= 5 (after normalization) in 90% of the samples of each group.
That would mean to filtering by group and the put all miRNAs together.

```{r sc}
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_fill_gradientn(colours = myPalette(100), limits=c(1, 20))

clean_table = function(what, mat, meta_clean, new_factors){
    i = meta_clean %>% filter(Diffname_short %in% what)
    m = mat[,i$UID]
    c = do.call(paste0,i[,new_factors])

    d = data.frame(row.names = colnames(m), cov=c)
    dds = DESeqDataSetFromMatrix(m,colData = d,design = ~ 1)
    dds = estimateSizeFactors(dds)
    mat_norm = counts(dds, normalized=T)
    cutoff = round(ncol(mat_norm)*0.9)
    clean = as.data.frame((mat_norm[rowSums(mat_norm>5)>cutoff,]))
    list(mat=clean, meta=i)
}

plot_density = function(clean, i){
    
    #voom_mat = voom(mat_norm, design = design, plot=T,lib.size=size_factor)
    #corrected = voom_mat$E / voom_mat$weight
    
    x = melt(log2(clean+1) %>% mutate(mir=rownames(clean)))
    idx = match(x$variable, i$UID)
    x$type = i[idx,"Diffname_short"]
    x$x_jitter = 1
    print( ggplot(x %>% filter(value>0), aes(x=value,group=variable,colour=type)) + geom_line(aes(color=type), stat="density",  alpha=0.4) + xlab("expression")+facet_wrap(~variable) )
    
    idx = sort(rowMeans(clean),index.return =TRUE)
    x$mir = factor(x$mir, levels=row.names(clean)[idx$ix])
    print(ggplot(x) + 
        geom_tile(aes(x=variable, y=mir, fill=value)) +
        sc +
        scale_x_discrete(labels="") +
        scale_y_discrete(labels="") )
}

# mat_clean =  (mat_3state[rowSums(mat_3state>5)>32,])
sc_clean = clean_table("SC",mat, meta_clean, new_factors)
plot_density(sc_clean[[1]], sc_clean[[2]])
```


## clustering of all groups after cleaning

```{r play2-other-group}

mesoearly_clean = clean_table("MESOEarly",mat, meta_clean, new_factors)
# plot_density(meso5_clean[[1]], meso5_clean[[2]])
mesolate_clean = clean_table("MESOLate",mat, meta_clean, new_factors)
ecto_clean = clean_table("ECTO",mat, meta_clean, new_factors)
# plot_density(ecto_clean[[1]], ecto_clean[[2]])
eb_clean = clean_table("EB",mat, meta_clean, new_factors)
de_clean = clean_table("DE",mat, meta_clean, new_factors)

mir_sub = unique(c(row.names(sc_clean$mat),
                   row.names(ecto_clean$mat),
                   row.names(eb_clean$mat),
                   row.names(de_clean$mat),
                   row.names(mesolate_clean$mat),
                   row.names(mesoearly_clean$mat)))
samples_sub = unique(c(colnames(sc_clean$mat),
                       colnames(ecto_clean$mat),
                       colnames(eb_clean$mat),
                       colnames(de_clean$mat),
                       colnames(mesolate_clean$mat),
                       colnames(mesoearly_clean$mat)))

new_factors = c("Diffname_short", "Tissue_of_Origin","run","Originating_Lab_ID")

mat_sub = mat[mir_sub, samples_sub]
meta_sub = meta_clean[samples_sub,]
covariates = do.call(paste0,meta_sub[,new_factors])
qc_figures(mat_sub, meta_sub, covariates)
```

### heatmap with voom normalization

```{r voom-norm}
voom_mat = voom(mat_sub, design = model.matrix(~ 0 + ., model.frame(~ 0 + . , meta_sub[,new_factors])), plot=T)
pheatmap(voom_mat$E, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type"),drop=F], show_rownames = F)
```

### heatmap with rlog-deseq normalization

```{r deseq-norm}
d = meta_sub[,new_factors]
dds = DESeqDataSetFromMatrix(mat_sub,colData = d,design = ~ Diffname_short)
rlog_mat = rlog(dds)
rlog_Mat = assay(rlog_mat)
rlog_Mat[rlog_Mat > 20]=20
meta_sub$mirna_b = cut(meta_sub$mirna, breaks=c(0,6,6.6,9))
pheatmap(rlog_Mat, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna_b"),drop=F], show_rownames = F)
```

## Non negative matrix factorization

Using this package to find modules( sets of mirnas that explain the different groups.)

http://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf

```{r nmf}
library(NMF)
res_nmf = nmf(log2(rlog_Mat^2+2), 3:7)
consensusmap(res_nmf, annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])
plot(res_nmf)
res_nmf_4 = nmf(log2(rlog_Mat^2+2), 4)
coefmap(res_nmf_4,annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])
basismap(res_nmf_4, subsetRow=TRUE)
consensusmap(res_nmf_4, annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])

# metagene
s = featureScore(res_nmf_4)
s = extractFeatures(res_nmf_4)
nmf_select = rownames(rlog_Mat[unique(unlist(s)),])
# X <- syntheticNMF(100, 3, 20, noise = 2)

pheatmap(rlog_Mat[unique(unlist(s)),], clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna_b"),drop=F], show_rownames = F)

```

number of differents miRNAs in modules `r length(unlist(s))`

## Differential expression

```{r limma}
library(edgeR)
meta_sub$Originating_Lab_ID = as.numeric(as.factor(meta_sub$Originating_Lab_ID))
meta_sub$run = as.numeric(as.factor(meta_sub$run))
meta_sub$Tissue_of_Origin = gsub(" ",".",meta_sub$Tissue_of_Origin)


dge <- DGEList(counts=mat_sub)
dge <- calcNormFactors(dge)
dm = getDesignMatrix(meta_sub[, new_factors], new_factors)
v <- voom(dge,dm$design,plot=TRUE)
fit <- lmFit(v,dm$design)
my.contrast = makeContrasts( ECTO.DE=Diffname_short.ECTO - Diffname_short.DE, 
                             ECTO.SC=Diffname_short.ECTO-Diffname_short.SC,
                             ECTO.EB=Diffname_short.ECTO-Diffname_short.EB,
                             ECTO.MESOEarly=Diffname_short.ECTO-Diffname_short.MESOEarly,
                             SC.DE=Diffname_short.SC - Diffname_short.DE, 
                             SC.EB=Diffname_short.SC-Diffname_short.EB,
                             SC.MESOEarly=Diffname_short.SC-Diffname_short.MESOEarly,
                             EB.DE=Diffname_short.EB - Diffname_short.DE, 
                             EB.MESOEarly=Diffname_short.EB-Diffname_short.MESOEarly,
                             MESOEarly.DE = Diffname_short.MESOEarly - Diffname_short.DE,
                             MESOEarly.MESOLate = Diffname_short.MESOEarly - Diffname_short.MESOLate,
                             MESOLate.DE = Diffname_short.MESOLate - Diffname_short.DE,
                             levels=dm$design)
fit = contrasts.fit(fit, my.contrast)
fit <- eBayes(fit)

top = list()
for (c in colnames(my.contrast)){
    top[[c]] = topTable(fit, coef=c,number=Inf) %>% mutate(mirs=rownames(fit$coefficients)) %>% filter(adj.P.Val<0.01, abs(logFC)>2)    
}

sapply(top, nrow)

# table(decideTests(fit,lfc = 2,p.value = 0.01))

```


unique miRNAs `r length(unique(unlist(sapply(top, function(x){x$mirs}))))`

```{r limma-heatmap}
select = as.character(unique(unlist(sapply(top, function(x){x$mirs}))))

pheatmap(rlog_Mat[select,], clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna_b"),drop=F], show_rownames = F)

```

```{r common}
intersect(nmf_select, select)
```

