
---
title: "Covariate Analysis for miRNA data with mixed effect modeling"
author: "Thanneer Perumal,Lorena Pantano"
date: "`r date()`"
output: html_document
---
```{r render, echo=FALSE, eval=FALSE}
library(rmarkdown)
render("CovariateAnalysis_miRNA_mixedEffects_All.Rmd")
```


```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
# library('ctv')
library('psych')
library('reshape2')
library('vcd')
# library('erer')
library('fpc')
# library('data.table')
library('dplyr')
library('knitr')
library('stringr')
#library('knit2synapse')
options(bitmapType = 'cairo')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")


synapseLogin()

# source utility files from ../R/lib folder
# file.sources = list.files('../R/lib',pattern="*.R")
# file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
# tmp = sapply(file.sources,source,.GlobalEnv)
source("../R/lib/calcResiduals.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/clean_mirna.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getDesignMatrix.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/linColumnFinder.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getGeneFilteredGeneExprMatrix.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/rownameToFirstColumn.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getAssociationStatistics.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getFactorAssociationStatistics.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getDesignMat2Fact.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/runPCAandPlotCorrelations.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/runPCA.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/calcCompleteCorAndPlot.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getFactorAssociationStatistics.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getFactorContAssociationStatistics.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/corMatFDRthreshFunc.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/plotCorWithCompare.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/designMatVarsToCovars.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getCovariatesMap.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/getSymbolicNamesList.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/designMat2CovEffects.R")

options(xtable.type="html")

knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=9,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='',
               bootstrap.show.code=FALSE, echo=FALSE)
```

### Download data

```{r setup, include=FALSE}
# Input Parameters

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Gender', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')
# ContCovariates = c('PassageAtThaw', 'PassageAtHarvest')
ContCovariates = NULL

```

Factor covariates considered for analysis are `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`.

Obtain count matrix and metadata from synapse.and remove samples with < 0.5 mill reads.

```{r load-clean}
# Get metadata
metadata_id = "syn3219876"
metadata_obj = synTableQuery(paste('SELECT * FROM',metadata_id,sep=' '))
metadata = metadata_obj@values %>% distinct(UID)
row.names(metadata) = metadata$UID

mat_id = "syn3355993"
mat_obj =  synGet(id=mat_id)
mat = read.table(mat_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)

reads_id = "syn3355994"
reads_obj =  synGet(id=reads_id)
reads = read.table(reads_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)

```


### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples with abnormal karyotypes.


```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
metadata_filtered <- 
  metadata %>%
  filter(Diffname_short != "") %>%
  filter(UID %in% colnames(mat)) %>%
  filter(Cell_Type == "PSC") %>%  
  filter(C4_Karyotype_Result != "abnormal")

REMOVED_UID <- setdiff(colnames(mat), metadata_filtered$UID)
metadata <- metadata[metadata_filtered$UID,]
mat <- mat[, metadata$UID]
```
The following `r length(REMOVED_UID)` samples were removed:

`r paste(gsub('_','\\\\_',REMOVED_UID), collapse= ',')` 

`r dim(mat)[1]` genes and `r dim(mat)[2]` samples are considered for the analysis

```{r data.split}
processed = clean_and_sync_mirna_tables(metadata, mat, reads, TRUE)
PROCESSED_COUNTS = processed$count
METADATA = processed$metadata

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
#COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```


### Covariates values

```{r covariates.values}
summary(COVARIATES)
```


### Covariate clustering
Determine relationship between covariates

Covariate correlation

```{r covariates.correlation, fig.width=10, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)

```{r initial.voom.normalisation}
# Initial normalisation of gene expression
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)

# Initial design matrix with covariates under consideration
DM = getDesignMatrix(COVARIATES, Intercept = F)
```

Expression distribution

```{r expr.distribution}
dd = melt(VOOM.GENE_EXPRESSION$E)
ggplot(dd, aes(x=value, group=X2)) +
    geom_density() +
    theme_bw() +
    xlab("normalized voom expression")
```


Clustering of initial normalised data (with NULL design)

```{r decompse.normalise.data, fig.height=8, fig.width=8}
# Find principal components of expression to plot
PC <- prcomp(VOOM.GENE_EXPRESSION$E)

# Determine number of clusters automatically using pamk
pam.cluster <- pamk(t(VOOM.GENE_EXPRESSION$E), krange=2:10)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2], 
                       cluster=pam.cluster$pamobject$clustering)

plotdata <- merge(plotdata, METADATA, by="UID")
plotdata <- mutate(plotdata, labels = '');#ifelse(Diffname_short == '',as.character(UID),'')) 

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=factor(cluster), shape=Diffname_short, size=Gender))
p <- p + theme_bw() + scale_size_manual(values = c(4, 2)) + scale_shape_manual(values=1:7) + theme(legend.position="top") 
p <- p + geom_text(aes(label= labels), size=4, hjust=0)
p
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates

```{r preAdjusted.covariates}
# Find correlation between PC's of gene expression with covariates
DESIGN = getDesignMat2Fact(DM$design,FactorCovariates)
colnames(DESIGN) = substr(colnames(DESIGN),1,50)
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DESIGN,'NULL design(voom-normalized)', isKeyPlot=TRUE)

# Find significant covariates
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),preAdjustedSigCovars$significantCovars)
```
Significant covariates to adjust at FDR 0.1 are `r paste(gsub('_','\\\\_',adjustCovars), collapse= ',')`

```{r preAdjustedSigCovars.NULL, fig.width=25, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r preAdjustedSigCovars.NULL.ALL, fig.width=25, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (iterative)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix. Here Donor_ID is chosen as random effect

```{r iterative.adjusted.voom.normalisation, cache=TRUE}
postAdjustCovars = list()
postAdjustCovars$fixed = ''
postAdjustCovars$random = 'Donor_ID'
 
# Assign residual covariates
residualSigCovars = preAdjustedSigCovars

significantCovars = designMatVarsToCovars(getCovariatesMap(DM), residualSigCovars$significantCovars)
covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),residualSigCovars$Effects.significantCovars)
covariatesEffects = covariatesEffects[setdiff(significantCovars,c(postAdjustCovars$fixed,postAdjustCovars$random))]

postAdjustCovars$fixed = names(which.max(covariatesEffects))

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 100){
  print(paste('Using following covariates in the model',
              paste(paste(postAdjustCovars$fixed,collapse=','),'as fixed effects and', paste(postAdjustCovars$random,collapse=','),'as random effects'),
              sep=':'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars$fixed,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
  # Estimate correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
  correlation <- duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, block=COVARIATES$Donor_ID)
  
  # Re-calculate voom weights with correlation of random effects
  VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F,
                                       block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights,
              block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars$fixed)
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                dplyr::select(DESIGN,-starts_with('Donor_ID')),
                                                'all adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  significantCovars = designMatVarsToCovars(getCovariatesMap(DM), residualSigCovars$significantCovars)
  covariatesEffects = designMat2CovEffects(getCovariatesMap(DM),residualSigCovars$Effects.significantCovars)
  covariatesEffects = covariatesEffects[setdiff(significantCovars,c(postAdjustCovars$fixed,postAdjustCovars$random))]
  
  postAdjustCovars$fixed = c(postAdjustCovars$fixed,names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}
tmp <- paste('Using following covariates in the final model', paste(paste(gsub('_','\\\\_',postAdjustCovars$fixed),collapse=','),'as fixed effects and', paste(gsub('_','\\\\_',postAdjustCovars$random),collapse=','),'as random effects'))
```

`r tmp`

Degenerate variables are: 

```{r adjusted.voom.normalisation}
# Post adjusted design matrix
DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars$fixed,drop=F], Intercept = F)
tmp = linColumnFinder(DM1$design)
print(tmp$relations)
DM1$design = DM1$design[,tmp$indepCols] 

# Recalculate weights with new design
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
correlation <- duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, block=COVARIATES$Donor_ID)

# Re-calculate voom weights with correlation of random effects
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=T,
                                     block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Linear model with new weights and design
VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights, block=COVARIATES$Donor_ID, correlation = correlation$cor)  
```

### Sanity Check
Residualise significant covariates using a linear model and find correlation between PCA of residuals with covariates

```{r calculate.residuals, include=FALSE}
# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                              dplyr::select(DESIGN,-starts_with('Donor_ID')),
                                              'residual matrix of all adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```

```{r residualSigCovars, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[1]]$plotData
```

```{r residualSigCovars.ALL, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```

### Normalisation (custom/manual design)

Reperform normalisation with custom design matrix (manually with Diffname short, Gender, Cell Type of Origin Level2, Donor Life Stage, lane, -Culture Conditions-, run as fixed effects and Donor ID as random effects)

```{r reperform.normalisation}
# Post adjusted design matrix
DM1 = getDesignMatrix(dplyr::select(COVARIATES,Diffname_short,Gender,Cell_Type_of_Origin_Level2,Donor_Life_Stage,lane,Culture_Conditions,run),Intercept = F)
DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    
# Estimate correlation of random effects
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F)
correlation <- duplicateCorrelation(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, block=COVARIATES$Donor_ID)
  
# Re-calculate voom weights with correlation of random effects
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=F,
                                     block=COVARIATES$Donor_ID, correlation = correlation$cor)
  
# Re-calculate linear fit using new weights and new design
VOOM.ADJUSTED.FIT = lmFit(VOOM.ADJUSTED.GENE_EXPRESSION$E, design=DM1$design, weights = VOOM.ADJUSTED.GENE_EXPRESSION$weights,
                          block=COVARIATES$Donor_ID, correlation = correlation$cor)

# Residuals after normalisation
RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,VOOM.ADJUSTED.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION,
                                              dplyr::select(DESIGN,-starts_with('Donor_ID')),
                                              'all adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)

# Residualise matrix and add back differentiation stages
RESID.EXPR = calcResiduals(geneBySampleValues = VOOM.ADJUSTED.GENE_EXPRESSION$E, 
                           samplesByCovariates = VOOM.ADJUSTED.GENE_EXPRESSION$design, 
                           varsToAddBackIn = c("Diffname_shortDE", "Diffname_shortEB", "Diffname_shortECTO", "Diffname_shortMESOLATE", "Diffname_shortMESO5", "Diffname_shortSC"), 
                           sampleWeights = VOOM.ADJUSTED.GENE_EXPRESSION$weights)
```

```{r residualSigCovars.manual, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[1]]$plotData
```

```{r residualSigCovars.manual.ALL, fig.width=25, fig.height=12}
residualSigCovars[["PC_res"]][[2]]$plotData
```

PCA of adjusted expression

```{r new.design, cache=TRUE, fig.height=8, fig.width=8}
# Find principal components of expression to plot
PC <- prcomp(RESID.EXPR)

# Determine number of clusters automatically using pamk
pam.cluster <- pamk(t(RESID.EXPR), krange=2:10)

# Plot first 2 PCs
plotdata <- data.frame(UID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC3=PC$rotation[,3], 
                       cluster=pam.cluster$pamobject$clustering)

plotdata <- mutate(plotdata, labels = "") 

plotdata <- merge(plotdata, METADATA, by="UID")

p <- ggplot(plotdata, aes(x=PC1, y=PC3))
p <- p + geom_point(aes(color=factor(cluster), shape=Diffname_short, size=Gender))
p <- p + theme_bw() + scale_size_manual(values = c(4, 2)) + scale_shape_manual(values=1:7) + theme(legend.position="top") 
p <- p + geom_text(aes(label= labels), size=4, hjust=0)
p
```

### Synapse Store
Store processed cpm counts, covariates, design matrix, and voom adjusted gene expression matrices

```{r synapsestore, echo=FALSE, include=FALSE, eval=FALSE}
parentId = "syn4595971"

ActivityName <- 'Covariate Analysis of miRNA data with mixed effects modeling'
  
thisFileName <- 'CovariateAnalysis_miRNA_mixedEffects_All.Rmd'
  
# Github link
library(rGithubClient)
thisRepo <- getRepo(repository = "lpantano/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='master')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Populate wiki with results
ALL_USED_IDs <- c(metadata_id, reads_id, mat_id)

CODE <- File(thisFileName,name = ActivityName,parentId = parentId)
CODE <- synStore(CODE, used = ALL_USED_IDs, activityName=activityName, executed=thisFile)

# Store counts matrix
COUNTS <- rownameToFirstColumn(VOOM.ADJUSTED.GENE_EXPRESSION$E,'GeneName')
write.table(COUNTS,file = '../../data/Processed_CPM_Counts.tsv',sep='\t',row.names=F,col.names=T,quote=F)
COUNTS <- File('../../data/Processed_CPM_Counts.tsv',name = 'Processed logCPM',parentId = parentId)
COUNTS <- synStore(COUNTS, used = ALL_USED_IDs, activityName = ActivityName, executed = CODE)  
  
# Adjusted Expression Data
RESID.EXPR <- rownameToFirstColumn(RESID.EXPR,'GeneName')  
write.table(RESID.EXPR,file = '../../data/Adjusted_miRNA_Expression.tsv',sep='\t',row.names=F,col.names=T,quote=F)
EXPR <- File('../../data/Adjusted_miRNA_Expression.tsv',name = 'Adjusted Expression',parentId = parentId)
EXPR <- synStore(EXPR, used = ALL_USED_IDs, activityName = ActivityName, executed = CODE)  
  
# Adjustment Weights 
WEIGHTS <- cbind(RESID.EXPR$GeneName,VOOM.ADJUSTED.GENE_EXPRESSION$weights)
colnames(WEIGHTS) <- colnames(RESID.EXPR)  
write.table(WEIGHTS,file = '../../data/Adjusted_miRNA_Weights.tsv',sep='\t',row.names=F,col.names=T,quote=F)
WEIGHTS <- File('../../data/Adjusted_miRNA_Weights.tsv',name = 'Adjusted Weights',parentId = parentId)
WEIGHTS <- synStore(WEIGHTS, used = ALL_USED_IDs, activityName = ActivityName, executed = CODE)
  
# Covariates Matrix
COVARIATES <- rownameToFirstColumn(COVARIATES,'UID')
write.table(COVARIATES,file = '../../data/Adjusted_miRNA_Covariates.tsv',sep='\t',row.names=F,col.names=T,quote=F)
COVARIATES <- File('../../data/Adjusted_miRNA_Covariates.tsv',name = 'Adjusted Covariates',parentId = parentId)
COVARIATES <- synStore(COVARIATES, used = ALL_USED_IDs, activityName = ActivityName, executed = CODE)

# Design Matrix
DESIGN <- rownameToFirstColumn(VOOM.ADJUSTED.GENE_EXPRESSION$design,'UID')
write.table(DESIGN,file = '../../data/Adjusted_miRNA_Design.tsv',sep='\t',row.names=F,col.names=T,quote=F)
DESIGN <- File('../../data/Adjusted_miRNA_Design.tsv',name = 'Adjusted Design',parentId = parentId)
DESIGN <- synStore(DESIGN, used = ALL_USED_IDs, activityName = ActivityName, executed = CODE)

devtools::source_gist("6117476")
knit2synapse("~/repos/pcbc_c4_analysis/code/Rmd/CovariateAnalysis_miRNA_mixedEffects_All.Rmd",
             owner=CODE@properties$id,
             overwrite=TRUE)
```
|  *Results*           |  *SynapseID*                                                               |
|  -------             |   ---------                                                                |
|  Adjusted Expression |  `r paste(EXPR$properties$id,EXPR$properties$versionNumber,sep='.')`       |

### Executed: Source code
`r paste0('[Source R Markdown](',print(thisFile),')')`
