---
title: Covariate Analysis for miRNA data with mixed effect modeling performed at indivudal
  differentiation stages
author: "Thanneer Perumal,Lorena Pantano"
date: '`r date()`'
output: html_document
---

```{r knitr,eval=F,echo=F}
library(rmarkdown)
# render("~/repos/pcbc_c4_analysis/code/Rmd/CovariateAnalysis_miRNA_MixedEffects_EachDiffState.Rmd")
```

```{r knitr-render, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=9,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='',
               bootstrap.show.code=FALSE, echo=FALSE)
```


```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file is

# Clear R console screen output
cat("\014")  

# Load required libraries
library('synapseClient')
library('RColorBrewer')
library('ggplot2')
library('gplots')
library('limma')
library('edgeR')
library('DESeq2')
# library('ctv')
library('psych')
library('reshape2')
library('vcd')
# library('erer')
# library('fpc')
# library('data.table')
library('dplyr')
library('knitr')
library('stringr')
#library('knit2synapse')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")


synapseLogin()

# source utility files from ../R/lib folder
# file.sources = list.files('../R/lib',pattern="*.R")
# file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
# tmp = sapply(file.sources,source,.GlobalEnv)
source("../R/lib/clean_mirna.R")
source("../R/lib/getDesignMatrix.R")
source("../R/lib/linColumnFinder.R")
source("../R/lib/getGeneFilteredGeneExprMatrix.R")
source("../R/lib/rownameToFirstColumn.R")
source("../R/lib/getAssociationStatistics.R")
source("../R/lib/getFactorAssociationStatistics.R")
source("../R/lib/getDesignMat2Fact.R")
source("../R/lib/runPCAandPlotCorrelations.R")
source("../R/lib/runPCA.R")
source("../R/lib/calcCompleteCorAndPlot.R")
source("../R/lib/getFactorAssociationStatistics.R")
source("../R/lib/getFactorContAssociationStatistics.R")
source("../R/lib/corMatFDRthreshFunc.R")
source("../R/lib/plotCorWithCompare.R")
source("../R/lib/designMatVarsToCovars.R")
source("../R/lib/getCovariatesMap.R")
source("../R/lib/getSymbolicNamesList.R")
source("../R/lib/designMat2CovEffects.R")

options(xtable.type="html")

```

### Download data

```{r setup, include=FALSE}
# Input Parameters

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short', 'run', 'lane', 'Cell_Line_Type', 'Cell_Line_of_Origin', 'Tissue_of_Origin', 'Reprogramming_Gene_Combination', 'Culture_Conditions', 'Donor_Life_Stage', 'Gender', 'Originating_Lab', 'Donor_ID', 'Cell_Type_of_Origin_Level2', 'Reprogramming_Vector_Type')
# ContCovariates = c('PassageAtThaw', 'PassageAtHarvest')
ContCovariates = NULL

```

Factor covariates considered for analysis are `r paste(gsub('_','\\\\_',FactorCovariates),collapse=',')`, and continuous covariates considered for the analysis are `r paste(gsub('_','\\\\_',ContCovariates),collapse=',')`.

Obtain count matrix and metadata from synapse.and remove samples with < 0.5 mill reads.

```{r load-clean}
# Get metadata
metadata_id = "syn3219876"
metadata_obj = synTableQuery(paste('SELECT * FROM',metadata_id,sep=' '))
metadata = metadata_obj@values %>% distinct(UID)
row.names(metadata) = metadata$UID

mat_id = "syn5014443"
mat_obj =  synGet(id=mat_id)
mat = read.table(mat_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)

reads_id = "syn5014445"
reads_obj =  synGet(id=reads_id)
reads = read.table(reads_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)

```


### Preprocess data
* Remove somatic samples and samples with not type.
* Remove samples with abnormal karyotypes.
* Remove samples < 500000 reads
* Remove samples < 20% mapping to miRNAs


```{r filtering, echo=TRUE}
#### Pre processing mRNA expression counts and metadata ####
processed = clean_and_sync_mirna_tables(metadata, mat, reads)
PROCESSED_COUNTS = processed$count
METADATA = processed$metadata
REMOVED_UID = setdiff(metadata$UID, METADATA$UID)
```
The following `r length(REMOVED_UID)` samples were removed:

`r paste(gsub('_','\\\\_',REMOVED_UID), collapse= ',')` 


```{r data.split}
# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates)]

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
#COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)

COVARIATES = split(COVARIATES,COVARIATES$Diffname_short)
```


Number of samples and miRNAs per condition.

```{r stats-samples-mirnas, results="asis"}
dd = sapply(names(PROCESSED_COUNTS), function(x){dim(PROCESSED_COUNTS[[x]]$filteredExprMatrix)})
rownames(dd) = c("mirnas", "samples")
kable(dd)
```


### Covariate clustering
Covariate correlation

```{r covariates.correlation, fig.width=10, fig.height=10}
covariateCorrelation <- function(COVARIATES){  
  Diffname = unique(COVARIATES$Diffname_short)
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Differentiation state:',as.character(Diffname)))
  writeLines('------------------------------------------------------------------------------')
  
  COVARIATES = droplevels(COVARIATES)
  COVARIATES = COVARIATES[,sapply(COVARIATES, function(x) length(levels(x))) > 1]
  
  # Correlation between covariates
  COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.1)  
  ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
  
  return(list(COVARIATES.CORRELATION = COVARIATES.CORRELATION))
}

COVARIATES.CORRELATION = lapply(COVARIATES, covariateCorrelation)
```

### Normalisation (with NULL)
Initial normalisation usign voom (with NULL design)

```{r initial.voom.normalisation, cache=FALSE}
initialVoomNormalisation <- function(COVARIATE, PROCESSED_COUNTS){
  writeLines('------------------------------------------------------------------------------')
  diffstate = as.character(unique(COVARIATE$Diffname_short))
  writeLines(paste('Differentiation state:',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')
  # Drop unilevel covariates
  COVARIATE = droplevels(COVARIATE)
  COVARIATE = COVARIATE[,sapply(COVARIATE, function(x) length(levels(x))) > 1]
  
  # Initial normalisation of gene expression
  # GENE.EXPRESSION = voom(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, design=NULL, plot=T)
  dds = DESeqDataSetFromMatrix(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, colData = COVARIATE, design = ~ 1)
  dds = DESeq(dds)
  GENE.EXPRESSION = assay(rlog(dds))
  plotDispEsts(dds)
  # Initial design matrix with covariates under consideration  
  DM = getDesignMatrix(COVARIATE, Intercept = F)
  
  return(list(GENE.EXPRESSION = GENE.EXPRESSION,
              COVARIATES = COVARIATE,
              DM = DM))
}
VOOM.GENE.EXPRESSION.PREADJUSTED = lapply(COVARIATES, initialVoomNormalisation, PROCESSED_COUNTS)
```

Clustering of initial normalised data (with NULL design)

```{r decompse.normalise.data, fig.height=8, fig.width=8}
plotPCAGeneExpression <- function(COVARIATES, VOOM.GENE.EXPRESSION.PREADJUSTED){ 
  writeLines('------------------------------------------------------------------------------')
  diffstate = as.character(unique(COVARIATES$Diffname_short))
  writeLines(paste('Differentiation state:',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')
  
  # Drop unilevel covariates
  COVARIATES = droplevels(COVARIATES)
  COVARIATES = COVARIATES[,sapply(COVARIATES, function(x) length(levels(x))) > 1]
  
  # Find principal components of expression to plot
  PC <- prcomp(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$GENE.EXPRESSION)
  
  # Plot first 2 PCs
  plotdata <- data.frame(UID=rownames(PC$rotation), 
                         PC1=PC$rotation[,1], 
                         PC2=PC$rotation[,2])

  plotdata <- merge(plotdata, METADATA, by="UID")
  plotdata <- mutate(plotdata, labels = '');#ifelse(Diffname_short == '',as.character(UID),'')) 

  p <- ggplot(plotdata, aes(x=PC1, y=PC2))
  p <- p + geom_point(aes(color=Cell_Line_Type, shape=Reprogramming_Vector_Type, size=Gender))
  p <- p + theme_bw() + scale_size_manual(values = c(4, 2)) 
  #+ scale_shape_manual(values=1:length(unique(COVARIATES$Reprogramming_Vector_Type))) + theme(legend.position="top") 
  p <- p + geom_text(aes(label= labels), size=4, hjust=0)
  print(p)
  
  return(list(p = p))
}
PCA.GENE.EXPRESSION = lapply(COVARIATES, plotPCAGeneExpression, VOOM.GENE.EXPRESSION.PREADJUSTED)
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates

```{r preAdjusted.covariates, fig.width=25, fig.height=12, cache=FALSE}
plotGeneCovariatesCorrelation <- function(COVARIATES, VOOM.GENE.EXPRESSION.PREADJUSTED){
  writeLines('------------------------------------------------------------------------------')
  diffstate = as.character(unique(COVARIATES$Diffname_short))
  writeLines(paste('Differentiation state:',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')
  
  # Drop unilevel covariates
  COVARIATES = droplevels(COVARIATES)
  COVARIATES = COVARIATES[,sapply(COVARIATES, function(x) length(levels(x))) > 1]
  
  # Find correlation between PC's of gene expression with covariates
  DESIGN = getDesignMat2Fact(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM$design,FactorCovariates)
  colnames(DESIGN) = substr(colnames(DESIGN),1,50)
  preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$GENE.EXPRESSION, DESIGN,'NULL design(voom-normalized)', isKeyPlot=TRUE)
  
  # Plot preAdjustedSigCovars
  print(preAdjustedSigCovars[["PC_res"]][[2]]$plotData)  
  
  # Find significant covariates
  adjustCovars = designMatVarsToCovars(getCovariatesMap(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM),
                                       preAdjustedSigCovars$significantCovars)
  
  writeLines('------------------------------------------------------------------------------')
  writeLines(paste('Significant covariates to adjust at FDR 0.1 are',paste(gsub('_','\\\\_',adjustCovars), collapse= ',')))
  writeLines('------------------------------------------------------------------------------')
  writeLines('')
  writeLines('')
  
  return(list(adjustCovars = adjustCovars,
              preAdjustedSigCovars = preAdjustedSigCovars))  
}
PREADJUSTED.COVARIATES = lapply(COVARIATES, plotGeneCovariatesCorrelation, VOOM.GENE.EXPRESSION.PREADJUSTED)
```

### Normalisation (iterative)
Since many covariates are correlated, re-normalising COVARIATES with an iterative design matrix. All effects are considered as fixed effects, except Donor_ID which is a random effect

```{r iterative.adjusted.voom.normalisation, fig.width=25, fig.height=12, cache=FALSE}
iterativeNormalisation <- function(COVARIATES, PROCESSED_COUNTS, VOOM.GENE.EXPRESSION.PREADJUSTED, PREADJUSTED.COVARIATES){  
  writeLines('------------------------------------------------------------------------------')
  diffstate = as.character(unique(COVARIATES$Diffname_short))
  writeLines(paste('Differentiation state:',as.character(diffstate)))
  writeLines('------------------------------------------------------------------------------')
  
  # Drop unilevel covariates
  COVARIATES = droplevels(COVARIATES)
  COVARIATES = COVARIATES[,sapply(COVARIATES, function(x) length(levels(x))) > 1]    
  
  # Assign return values
  FIT = c()
  CORRELATION = c()
  RESIDUAL.GENE.EXPRESSION = c()  
  postAdjustCovars = c()
  
  # Assign residual covariates
  residualSigCovars = PREADJUSTED.COVARIATES[[diffstate]]$preAdjustedSigCovars
  
  significantCovars = designMatVarsToCovars(getCovariatesMap(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM),
                                            residualSigCovars$significantCovars)
  covariatesEffects = designMat2CovEffects(getCovariatesMap(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM),
                                           residualSigCovars$Effects.significantCovars)
  covariatesEffects = covariatesEffects[setdiff(significantCovars,c(postAdjustCovars,'Donor_ID'))]
  # print(c('Init: ',covariatesEffects,'\n'))
  postAdjustCovars = names(which.max(covariatesEffects))
  DESIGN = dplyr::select(as.data.frame(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM$design),-starts_with('Donor_ID'))
  colnames(DESIGN) = substr(colnames(DESIGN),1,50)
  loopCount = 0 
  while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 2){
    writeLines(paste('Using following covariates in the model:',
                paste(paste(gsub('_','\\\\_',postAdjustCovars),collapse=', '),'as fixed effects'),
                sep=':'))
    
    # Post adjusted design matrix
    DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
    DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    # Estimate sample weights with null design
    GENE.EXPRESSION = voom(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, 
                           design = DM1$design, 
                           plot=F)
    # Estimate correlation between donor effects
    CORRELATION = duplicateCorrelation(GENE.EXPRESSION, 
                                       design = DM1$design,
                                       block = COVARIATES$Donor_ID)
    num_donor = unlist(table(COVARIATES$Donor_ID))
    if((CORRELATION$consensus.cor) <= 0 || abs(CORRELATION$consensus.cor) == 1
       # || sum(num_donor[num_donor!=1]) + 2 < ncol(DM1$design) 
       ){
      writeLines('Random effects correlation cannot be estimated')
      break
    }
    # Estimate sample weights with correlation efects
    GENE.EXPRESSION = voom(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, 
                           design = DM1$design,
                           #block = COVARIATES$Donor_ID,
                           #correlation = CORRELATION$consensus.cor,
                           plot=F)
    
    # Fit linear model with chosen design
    FIT = lmFit(GENE.EXPRESSION,
                block = COVARIATES$Donor_ID,
                correlation = CORRELATION$consensus.cor)
    
    # Residuals after normalisation
    RESIDUAL.GENE.EXPRESSION = residuals.MArrayLM(FIT,GENE.EXPRESSION$E)
    
    # Residual covariates to choose from
    residCovars <- setdiff(c(FactorCovariates,ContCovariates), postAdjustCovars)
    
    # Find PC of residual gene expression and significant covariates that are highly correlated with PCs 
    residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE.EXPRESSION, 
                                                  DESIGN,
                                                  'all adjusted design(voom-normalized)',
                                                  isKeyPlot=TRUE)
    
    # Add postadjusted covariates (if any)
    significantCovars = designMatVarsToCovars(getCovariatesMap(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM), 
                                              residualSigCovars$significantCovars)
    covariatesEffects = designMat2CovEffects(getCovariatesMap(VOOM.GENE.EXPRESSION.PREADJUSTED[[diffstate]]$DM),
                                             residualSigCovars$Effects.significantCovars)
    covariatesEffects = covariatesEffects[setdiff(significantCovars,c(postAdjustCovars,'Donor_ID'))]
    
    # print(c("res after loop ",residualSigCovars$significantCovars, "\n"))
    # print(colnames(VOOM.GENE.EXPRESSION.POSTADJUSTED$SC$GENE.EXPRESSION$design))
    postAdjustCovars = c(postAdjustCovars,names(which.max(covariatesEffects)))
    
    write.table(diffstate, file = './temp.txt', append=T)
    write.table(postAdjustCovars, file = './temp.txt', append=T)
    loopCount = loopCount + 1
  }
  
  
  writeLines('--------------------------------------------------------------------------------------------')
  writeLines(paste('Using following covariates in the final model', paste(paste(gsub('_','\\\\_',postAdjustCovars),collapse=','),'as fixed effects')))
  writeLines('--------------------------------------------------------------------------------------------')
  writeLines('')
  
  ### Sanity Check
  print(residualSigCovars[["PC_res"]][[2]]$plotData)
  writeLines('---------------------------------------------------------------------------------------------')
  if (length(postAdjustCovars)>0){
        # Post adjusted design matrix
    DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
    DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
    DM1$design = DM1$design[,colSums(DM1$design) > 1]
    # Estimate sample weights with null design
    GENE.EXPRESSION = voom(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, 
                           design = DM1$design, 
                           plot=F)
    # Estimate correlation between donor effects
    CORRELATION = duplicateCorrelation(GENE.EXPRESSION, 
                                       design = DM1$design,
                                       block = COVARIATES$Donor_ID)
  
    # Estimate sample weights with correlation efects
    GENE.EXPRESSION = voom(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix, 
                           design = DM1$design,
                           block = COVARIATES$Donor_ID,
                           correlation = CORRELATION$consensus.cor,
                           plot=F)
    
    # Fit linear model with chosen design
    FIT = lmFit(GENE.EXPRESSION,
                block = COVARIATES$Donor_ID,
                correlation = CORRELATION$consensus.cor)
    
    # Residuals after normalisation
    RESIDUAL.GENE.EXPRESSION = residuals.MArrayLM(FIT,GENE.EXPRESSION$E)
  
    return(list(GENE.EXPRESSION = GENE.EXPRESSION,
                FIT = FIT,
                RESIDUAL.GENE.EXPRESSION = RESIDUAL.GENE.EXPRESSION,
                residualSigCovars = residualSigCovars,
                postAdjustCovars = postAdjustCovars))
  }
}  

VOOM.GENE.EXPRESSION.POSTADJUSTED = lapply(COVARIATES, iterativeNormalisation, PROCESSED_COUNTS, VOOM.GENE.EXPRESSION.PREADJUSTED, PREADJUSTED.COVARIATES)

```

```{r show-postadjusted}
sapply(names(VOOM.GENE.EXPRESSION.POSTADJUSTED), function(x) VOOM.GENE.EXPRESSION.POSTADJUSTED[[x]]$postAdjustCovars)
```


### Synapse Store
Store processed cpm counts, covariates, design matrix, and voom adjusted gene expression matrices

```{r synapsestore, echo=FALSE, include=FALSE, cache=FALSE}
parentId = "syn5014476"
activityName <- 'Covariate Analysis of miRNA data at each diff state with mixed effects'
  
thisFileName <- 'CovariateAnalysis_miRNA_MixedEffects_EachDiffState.Rmd'
  
# Github link
library(rGithubClient)
thisRepo <- getRepo(repository = "lpantano/pcbc_c4_analysis", 
                    ref="branch", 
                    refName='master')

thisFile <- getPermlink(repository = thisRepo,
                        repositoryPath=paste0('code/Rmd/', thisFileName))
    
# Populate wiki with results
ALL_USED_IDs <- c(metadata_id, reads_id, mat_id)

CODE <- File(thisFileName,name = activityName,parentId = parentId)
CODE <- synStore(CODE, used = ALL_USED_IDs, activityName=activityName, executed=thisFile)

storeFilesInSynapse <- function(COVARIATES, PROCESSED_COUNTS, VOOM.GENE.EXPRESSION.POSTADJUSTED, CODE){
  diffstate = as.character(unique(COVARIATES$Diffname_short))
  print(diffstate)    
  # Drop unilevel covariates
  COVARIATES = droplevels(COVARIATES)
  COVARIATES = COVARIATES[,sapply(COVARIATES, function(x) length(levels(x))) > 1]    
  
  # Store counts matrix
  COUNTS <- rownameToFirstColumn(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix$counts,'GeneName')
  write.table(COUNTS,
              file = paste0('../../data/Processed_CPM_Counts_',diffstate,'.tsv'),
              sep='\t',
              row.names=F,
              col.names=T,
              quote=F)
  COUNTS_OBJ <- File(paste0('../../data/Processed_CPM_Counts_',diffstate,'.tsv'),
                     name = paste('Processed logCPM Counts',diffstate),
                     parentId = parentId)
  COUNTS_OBJ <- synStore(COUNTS_OBJ, 
                         used = ALL_USED_IDs, 
                         activityName = activityName, 
                         executed = CODE) 
  
  # Adjustment Weights
  WEIGHTS <- cbind(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix$genes,
                   VOOM.GENE.EXPRESSION.POSTADJUSTED[[diffstate]]$GENE.EXPRESSION$weights)
  colnames(WEIGHTS) <- c('GeneNames',rownames(PROCESSED_COUNTS[[diffstate]]$filteredExprMatrix$samples))
  write.table(WEIGHTS,
              file = paste0('../../data/Adjusted_miRNA_Weights_',diffstate,'.tsv'),
              sep='\t',
              row.names=F,
              col.names=T,
              quote=F)
  WEIGHTS_OBJ <- File(paste0('../../data/Adjusted_miRNA_Weights_',diffstate,'.tsv'),
                      name = paste('Adjusted Weights',diffstate),
                      parentId = parentId)
  WEIGHTS_OBJ <- synStore(WEIGHTS_OBJ, 
                          used = ALL_USED_IDs, 
                          activityName = activityName, 
                          executed = CODE)
  
  # Covariates Matrix
  needed = VOOM.GENE.EXPRESSION.POSTADJUSTED[[diffstate]]$postAdjustCovars
  
  COVARIATES <- rownameToFirstColumn(COVARIATES[,c(needed,"Donor_ID")],'UID')
  write.table(COVARIATES,
              file = paste0('../../data/Adjusted_miRNA_Covariates_',diffstate,'.tsv'),
              sep='\t',
              row.names=F,
              col.names=T,
              quote=F)
  COVARIATES_OBJ <- File(paste0('../../data/Adjusted_miRNA_Covariates_',diffstate,'.tsv'),
                         name = paste('Adjusted Covariates',diffstate),
                         parentId = parentId)
  COVARIATES_OBJ <- synStore(COVARIATES_OBJ, 
                             used = ALL_USED_IDs, 
                             activityName = activityName, 
                             executed = CODE)
  
  # Design Matrix
  DESIGN <- rownameToFirstColumn(VOOM.GENE.EXPRESSION.POSTADJUSTED[[diffstate]]$GENE.EXPRESSION$design,'UID')
  write.table(DESIGN,
              file = paste0('../../data/Adjusted_miRNA_Design_',diffstate,'.tsv'),
              sep='\t',
              row.names=F,
              col.names=T,
              quote=F)
  DESIGN_OBJ <- File(paste0('../../data/Adjusted_miRNA_Design_',diffstate,'.tsv'),
                     name = paste('Adjusted Design',diffstate),
                     parentId = parentId)
  DESIGN_OBJ <- synStore(DESIGN_OBJ,
                         used = ALL_USED_IDs, 
                         activityName = activityName, 
                         executed = CODE)
}

STORED.FILES = lapply(COVARIATES[! names(COVARIATES) %in% c('MESO30', 'MESO15')], storeFilesInSynapse, PROCESSED_COUNTS, VOOM.GENE.EXPRESSION.POSTADJUSTED, CODE)
```

```{r render-synapse, echo=FALSE, eval=FALSE}
devtools::source_gist("6117476")
knit2synapse("~/repos/pcbc_c4_analysis/code/Rmd/CovariateAnalysis_miRNA_MixedEffects_EachDiffState.Rmd",
             owner=CODE@properties$id,
             overwrite=TRUE)
```

