---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
```{r knitr,eval=F,echo=F}
library(rmarkdown)
library(knitrBootstrap)
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png", fig.width=9,fig.heigh=9,
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE, echo=FALSE)
render("~/repos/pcbc_c4_analysis/code/Rmd/DE_MatrixmiRNA_seqbuster.Rmd", output_dir = "~/Dropbox/Public/hsph/pcbc")
```

# metadata and covariates

```{r meta, cache=TRUE}
library(tidyr)
library(dplyr)
library(knitr)
library(synapseClient)
library(DESeq2)
library(edgeR)
library(vcd)
library(limma)
library(reshape2)
library(gplots)
library(ggplot2)
library(RColorBrewer)
library(corrplot)
library(psych)
library(pheatmap)

source("~/repos/pcbc_c4_analysis/code/R/lib/qc.R")
source("~/repos/pcbc_c4_analysis/code/R/lib/covariateAnalysis.R")

synapseLogin()

# Get metadata
metadata_id = "syn3219876"
metadata_obj = synTableQuery(paste('SELECT * FROM',metadata_id,sep=' '))
metadata = metadata_obj@values %>% distinct(UID)
row.names(metadata) = metadata$UID

mat_obj =  synGet(id="syn3355993")
mat = read.table(mat_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)
# mat = read.csv("~/repos/pcbc_c4_analysis/data/mirna.csv", check.names = FALSE)

reads_obj =  synGet(id="syn3355994")
reads = read.table(reads_obj@filePath, check.names = F, header=T, sep="\t", row.names=1)
# reads = read.table("~/repos/pcbc_c4_analysis/data/smallRNA_reads.tsv", check.names = FALSE, row.names=1, header=T)

keep = ( (reads %>% mutate(samples=rownames(reads)) %>% filter(mirna > 500000 & norm > 0.2))[,"samples"] )
metadata$size = reads[as.character(row.names(metadata)),"norm"]
metadata$size_cat = reads[as.character(row.names(metadata)),"breaks"]
metadata$mirna = log2(reads[as.character(row.names(metadata)),"mirna"])
# join meso-15 and meso-30
stages =  gsub("-5","_Early",gsub("-30","_Late",gsub("-15","_Late",metadata$Diffname_short)))
metadata$Diffname_short = as.character(stages)
```

## Clean and sync tables

Just make sure the tables are sync, same order for samples

```{r clean_metadata}
meta_clean = metadata[colnames(mat), ]
meta_clean[meta_clean == 'N/A'] = NA
meta_clean[is.na(meta_clean)] = 'unk'
meta_clean[meta_clean == ''] = 'unk'
```

```{r clean}
meta_clean = meta_clean[keep,]
mat = mat[,keep]

keep_meta = meta_clean$Diffname_short != "unk"
meta_clean = meta_clean[keep_meta,]
mat = mat[,keep_meta]

meta_clean$mirna= log10(colSums(mat))

new_factors = c("Diffname_short", "Tissue_of_Origin","run","Originating_Lab_ID")
```

# Filtering by group

I want to try if filtering by stable miRNAs give better signal. 
I keep miRNAs that are express >= 5 (after normalization) in 90% of the samples of each group.
That would mean to filtering by group and the put all miRNAs together.

```{r sc}
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
sc <- scale_fill_gradientn(colours = myPalette(100), limits=c(1, 20))

clean_table = function(what, mat, meta_clean, new_factors){
    i = meta_clean %>% filter(Diffname_short %in% what)
    m = mat[,i$UID]
    c = do.call(paste0,i[,new_factors])

    d = data.frame(row.names = colnames(m), cov=c)
    dds = DESeqDataSetFromMatrix(m,colData = d,design = ~ 1)
    dds = estimateSizeFactors(dds)
    mat_norm = counts(dds, normalized=T)
    cutoff = round(ncol(mat_norm)*0.9)
    clean = as.data.frame((mat_norm[rowSums(mat_norm>5)>cutoff,]))
    list(mat=clean, meta=i)
}

plot_density = function(clean, i){
    
    #voom_mat = voom(mat_norm, design = design, plot=T,lib.size=size_factor)
    #corrected = voom_mat$E / voom_mat$weight
    
    x = melt(log2(clean+1) %>% mutate(mir=rownames(clean)))
    idx = match(x$variable, i$UID)
    x$type = i[idx,"Diffname_short"]
    x$x_jitter = 1
    print( ggplot(x %>% filter(value>0), aes(x=value,group=variable,colour=type)) + geom_line(aes(color=type), stat="density",  alpha=0.4) + xlab("expression")+facet_wrap(~variable) )
    
    idx = sort(rowMeans(clean),index.return =TRUE)
    x$mir = factor(x$mir, levels=row.names(clean)[idx$ix])
    print(ggplot(x) + 
        geom_tile(aes(x=variable, y=mir, fill=value)) +
        sc +
        scale_x_discrete(labels="") +
        scale_y_discrete(labels="") )
}

sc_clean = clean_table("SC",mat, meta_clean, new_factors)
mesoearly_clean = clean_table("MESO_Early",mat, meta_clean, new_factors)
mesolate_clean = clean_table("MESO_Late",mat, meta_clean, new_factors)
ecto_clean = clean_table("ECTO",mat, meta_clean, new_factors)
eb_clean = clean_table("EB",mat, meta_clean, new_factors)
de_clean = clean_table("DE",mat, meta_clean, new_factors)

mir_sub = unique(c(row.names(sc_clean$mat),
                   row.names(ecto_clean$mat),
                   row.names(eb_clean$mat),
                   row.names(de_clean$mat),
                   row.names(mesolate_clean$mat),
                   row.names(mesoearly_clean$mat)))
samples_sub = unique(c(colnames(sc_clean$mat),
                       colnames(ecto_clean$mat),
                       colnames(eb_clean$mat),
                       colnames(de_clean$mat),
                       colnames(mesolate_clean$mat),
                       colnames(mesoearly_clean$mat)))


mat_sub = mat[mir_sub, samples_sub]
meta_sub = meta_clean[samples_sub,]
# covariates = do.call(paste0,meta_sub[,new_factors])
# qc_figures(mat_sub, meta_sub, covariates)
```

### heatmap with voom normalization

```{r voom-norm}
voom_mat = voom(mat_sub, design = model.matrix(~ 0 + ., model.frame(~ . , meta_sub[,new_factors])), plot=T)
pheatmap(voom_mat$E, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna"),drop=F], show_rownames = F)
```

### heatmap with rlog-deseq normalization

```{r deseq-norm, cache=TRUE}
d = meta_sub[,new_factors]
dds = DESeqDataSetFromMatrix(mat_sub,colData = d,design = ~ Diffname_short)
rlog_mat = rlog(dds)
rlog_Mat = assay(rlog_mat)
rlog_Mat[rlog_Mat > 20]=20
meta_sub$mirna_b = cut(meta_sub$mirna, breaks=c(0,6,6.6,9))
pheatmap(rlog_Mat, clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna"),drop=F], show_rownames = F)
```

## Non negative matrix factorization

Using this package to find modules( sets of mirnas that explain the different groups.)

http://cran.r-project.org/web/packages/NMF/vignettes/NMF-vignette.pdf

```{r nmf, cache=TRUE}
library(NMF)
# res_nmf = nmf(log2(rlog_Mat^2+2), 3:7)
# consensusmap(res_nmf, annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])
# plot(res_nmf)
res_nmf_4 = nmf(log2(rlog_Mat^2+2), 4, nrun=20)
coefmap(res_nmf_4,annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])
basismap(res_nmf_4, subsetRow=TRUE)
consensusmap(res_nmf_4, annCol = meta_sub[,c("Diffname_short","Cell_Line_Type")])

# metagene
s = featureScore(res_nmf_4)
s = extractFeatures(res_nmf_4)
nmf_select = rownames(rlog_Mat[unique(unlist(s)),])
# X <- syntheticNMF(100, 3, 20, noise = 2)

pheatmap(rlog_Mat[nmf_select,], clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna"),drop=F], show_rownames = F)
```

It worked without the MESOLate samples, no quite well detecting modules with all groups.

number of differents miRNAs in modules `r length(unlist(s))`

## Differential expression


```{r limma}

meta_sub$Originating_Lab_ID = as.numeric(as.factor(meta_sub$Originating_Lab_ID))
meta_sub$run = as.numeric(as.factor(meta_sub$run))
meta_sub$Tissue_of_Origin = gsub(" ",".",meta_sub$Tissue_of_Origin)

dge <- DGEList(counts=mat_sub)
dge <- calcNormFactors(dge)

dm = model.matrix(~ 0 + ., model.frame(~  . , meta_sub[,new_factors]))
v <- voom(dge,dm, plot=TRUE)
fit <- lmFit(v,dm)

contrasts_list <- combn(paste0("Diffname_short",unique(meta_sub$Diffname_short)),2)
contrasts_list <- apply(contrasts_list,2,paste,collapse='-')

contrast_matrix <- makeContrasts(contrasts=contrasts_list,levels=colnames(fit$coefficients))

# Refit contrasts
fit.cont <- contrasts.fit(fit,contrast_matrix)

# Estimate moderated t-statistics
fit.cont  <- eBayes(fit.cont )

# Obtain all the differential expession combinations
diff.exp <- list()
diff.exp$logFC <- data.frame(row.names = rownames(v$E))
diff.exp$adj.P.Val <- data.frame(row.names = rownames(v$E))

for (i in contrasts_list){
  tmp <- topTable(fit.cont, coef=i, number=Inf)
  colNames <- gsub('Diffname_short','',i)
  diff.exp$logFC[,colNames] <- tmp[rownames(diff.exp$logFC),'logFC']
  diff.exp$adj.P.Val[,colNames] <- tmp[rownames(diff.exp$adj.P.Val),'adj.P.Val']
}
tmp <- cbind(diff.exp$logFC, diff.exp$adj.P.Val)
# write.table(diff.exp,file='./DifferentialExpression_DiffnameShort.txt',sep='\t',row.names=T,col.names=T,quote=F)

# Get number of significantly differentialy expressed genes with adj.P.Val <= 0.01 and abs(logFC) >= 2
sig.exp <- diff.exp$adj.P.Val<=0.01 & abs(diff.exp$logFC) >= 2
# colnames(v) <- gsub('-adj.P.Val','',colnames(sig.exp))
sig.gset <- list()
for (i in colnames(sig.exp))
   sig.gset[[i]] <- rownames(sig.exp)[sig.exp[,i]]

# res = do.call(rbind, top)
write.table(res, "~/repos/pcbc_c4_analysis/data/miRNA_DifferentialExpression_DiffnameShort.txt", row.names=F, quote=F, sep="\t") 
```

Number of miRNAs de-regulated (abs(FC)>2 & adj.P.Val<0.01):

```{r show-de, results='asis'}
# kable(data.frame(sapply(top, nrow)))
print(apply(sig.exp,2,sum))
```

unique miRNAs `r length(as.character(unique(unlist(sig.gset))))`

```{r limma-heatmap}
select = as.character(unique(unlist(sig.gset)))

pheatmap(rlog_Mat[select,], clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna"),drop=F], show_rownames = F, show_colnames = F)
```

testing densort function

```{r dendsort}
library(dendsort)
par(mar=c(1,1,1,1), cex=0.5)
sideColors = rainbow(6)
names(sideColors) = unique(meta_sub$Diffname_short)
hc = dendsort(hclust(as.dist(1-cor(rlog_Mat[select,])^2), method="ward.D"))
hr = dendsort(hclust(as.dist(1-cor(t(rlog_Mat[select,]))^2), method="ward.D"))

myPalette <-  colorRampPalette(c("red", "white", "green"))(n = 98)
colors = c(seq(-2,5,length=33),seq(5,12,length=33),seq(12,21,length=33))


heatmap.2(rlog_Mat[select, ], dendrogram ="both",
          col=myPalette,breaks=colors,
          Rowv=(as.dendrogram(hr)), Colv=(as.dendrogram(hc)),
          scale="none", labRow="", labCol="", ColSideColors = sideColors[meta_sub$Diffname_short],
          key = F, trace="none", density.info="none")

legend("topleft", legend=names(sideColors), fill=sideColors, cex = 0.5, horiz = T)

#plot(1:6,pch=20, col=sideColors)
```


```{r common, echo=FALSE, eval=FALSE}
common = intersect(nmf_select, select)
pheatmap.2(rlog_Mat[common,], clustering_distance_rows = "correlation", clustering_distance_cols = "correlation", clustering_method = "ward", annotation = meta_sub[,c("Diffname_short","Cell_Line_Type","size","mirna"),drop=F], show_rownames = F)
```

```{r store,eval=F}
SYNAPSE_STORE = FALSE
parentId = "syn3464726"
if (SYNAPSE_STORE){
  COVARIATES_CODE <- File('DE_MatrixmiRNA_seqbuster.Rmd',name = 'miRNA Differential Expression Code',parentId = parentId)
  COVARIATES_CODE <- synStore(COVARIATES_CODE, used = c(mat_obj$properties$id,metadata_obj@schema),activityName='miRNA Differential Expression Code', executed='https://github.com/lpantano/pcbc_c4_analysis/blob/master/code/Rmd/DE_MatrixmiRNA_seqbuster.Rmd')
  DE <- File('"~/repos/pcbc_c4_analysis/data/DifferentialExpression_DiffnameShort.txt',name = 'Differential Expression DiffnameShort',parentId = parentId)
  DE <- synStore(DE, used = c(mat_obj$properties$id,metadata_obj@schema),activityName='miRNA Differential Analysis', executed=COVARIATES_CODE)

devtools::source_gist("6117476")
#knit2synapse("~/repos/pcbc_c4_analysis/code/Rmd/DE_MatrixmiRNA_seqbuster.Rmd",
#             owner="syn3356340",
#             overwrite=TRUE)

}
```

